src/debugger.rs

```diff
@@ -0,0 +1,611 @@
+                    },
+                    BpfRegId::Sp => "sp",
+                    BpfRegId::Pc => "pc",
+                    BpfRegId::InstructionCountRemaining => "remaining",
+                    _ => "unknown",
+                }
+                .into();
+                let set = String::from("General Purpose Registers");
+                let generic = match r {
+                    BpfRegId::Sp => Some(Generic::Sp),
+                    BpfRegId::Pc => Some(Generic::Pc),
+                    _ => None,
+                };
+                let reg = Register {
+                    name: &name,
+                    alt_name: None,
+                    bitsize: (usize::from(size)) * 8,
+                    offset: reg_id * (usize::from(size)),
+                    encoding: Encoding::Uint,
+                    format: Format::Hex,
+                    set: &set,
+                    gcc: None,
+                    dwarf: Some(reg_id),
+                    generic,
+                    container_regs: None,
+                    invalidate_regs: None,
+                };
+                Ok(reg_info.write(reg))
+            }
+        }
+    }
+}
+
+mod bpf_arch {
+    use gdbstub::arch::{Arch, SingleStepGdbBehavior};
+
+    /// BPF-specific breakpoint kinds.
+    ///
+    /// Extracted from the GDB source code [BPF Breakpoint Kinds](https://github.com/bminor/binutils-gdb/blob/9e0f6329352ab9c5e2f278181a3875918cff3b27/gdb/bpf-tdep.c#L205)
+    #[derive(Debug)]
+    pub enum BpfBreakpointKind {
+        /// BPF breakpoint
+        BpfBpKindBrkpt,
+    }
+
+    impl gdbstub::arch::BreakpointKind for BpfBreakpointKind {
+        fn from_usize(kind: usize) -> Option<Self> {
+            let kind = match kind {
+                0 => BpfBreakpointKind::BpfBpKindBrkpt,
+                _ => return None,
+            };
+            Some(kind)
+        }
+    }
+
+    /// Implements `Arch` for 64-bit BPF.
+    pub enum Bpf64 {}
+
+    #[allow(deprecated)]
+    impl Arch for Bpf64 {
+        type Usize = u64;
+        type Registers = reg::BpfRegs<u64>;
+        type RegId = reg::id::BpfRegId<u64>;
+        type BreakpointKind = BpfBreakpointKind;
+
+        #[inline(always)]
+        fn single_step_gdb_behavior() -> SingleStepGdbBehavior {
+            SingleStepGdbBehavior::Required
+        }
+    }
+
+    pub mod reg {
+        pub use bpf::BpfRegs;
+
+        mod bpf {
+            use core::convert::TryInto;
+
+            use num_traits::PrimInt;
+
+            use gdbstub::arch::Registers;
+            use gdbstub::internal::LeBytes;
+
+            /// BPF registers.
+            ///
+            /// The register width of the SP and PC register is set to `u32` or `u64` based on the `<U>` type.
+            ///
+            /// Source: <https://github.com/bminor/binutils-gdb/blob/9e0f6329352ab9c5e2f278181a3875918cff3b27/gdb/bpf-tdep.c#L42>
+            #[derive(Debug, Default, Clone, Eq, PartialEq)]
+            pub struct BpfRegs<U> {
```

Comment:
yes I'll switch those to plain `u64`

Ideally I would like to upstream the whole `bpf_arch` module at some point so it can be used from the [gdbstub_arch](https://github.com/daniel5151/gdbstub/tree/master/gdbstub_arch/src) crate directly. The thing is that normal GDB is using `u32` for SP and PC for the bpf arch, so that's why there is a generic here. But that shouldn't bother us for now. (If upstreamed, it would look something like [this](https://github.com/jawilk/gdbstub/blob/bpf/gdbstub_arch/src/bpf/mod.rs#L26-L64))